README
This package contains the C++ implementation of the 
Content-Based Networking Simulation library.
 
Copyright (C) 2001-2004 University of Colorado

Authors: Matthew J.  Rutherford <rutherfo@cs.colorado.edu>
	 Antonio Carzaniga <carzanig@cs.colorado.edu>
         See AUTHORS for full details.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License (file LICENSE) for more details.

You should have received a copy of the GNU General Public License
along with this program (see LICENSE); if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
USA, or send email to serl@cs.colorado.edu.

CONTACT

Please, address technical questions, bug reports and any sort of
technical and non-technical feedback to:

   Matthew J. Rutherford <rutherfo@cs.colorado.edu>

DOCUMENTATION

The documentation for this package is available on-line at
http://www.cs.colorado.edu/~carzanig/cbn

INSTALLATION

Instructions for the impatient:

  1. ./configure
  2. make

Please, read the INSTALL file for further installation instructions.


ChangeLog
2012-06-19  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* cbcbsim.cc and many other files
	workoad file does no more show the topology. Topoogy is passed when simulation is launched (Notice that now, in the brite file, the link 		propagation delay is exppressed in seconds!!!)
2011-11-14  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	*  CBCBDataCollector.cc and many other files
	Interest forwarding in line statistics
2011-11-01  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	*  CBCBNode.cc and many other files
	Interest forwarding strategies implemented, rnd and rnd balancing
2011-09-20  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc, CBCBNode.cc and many other files
	added the possibility to specify from the topo.brite file the caching strategy of a single node so that 2 nodes can have different replacement 		policies and some work on the code style
2011-06-16  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/SocketManager.cc include/SocketManager.h and many others
	improved AIMD, halve the cwin only if t_last_halve+T.O > t_new_TO. Pay attention to re-expression of interest (not needed at the moment)     expecially with the multipath
2011-06-16  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/SocketManager.cc include/SocketManager.h and many others
	PIT_timeout is now a receiver timeout, and it's decoupled with the real timeout of the PIT (need to change the name)
2011-06-09  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/SocketManager.cc include/SocketManager.h and many others
	Do not resend immediately an interest if a time out occoured (verify if work with multipath and modify send_data_interest in Diego's sim versions)
2011-05-18  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/SPTsBroadcastLayer.cc, src/CBCBNode.cc, src/CCNx*.cc
	Computing alternative paths( avoid FIB cycles by not forwarding INTEREST to ifid from which you received the interest ), exploiting of Carzaniga's routing table (forward the INTEREST as a message inverting the sense of the Carzaniga's routing protocol)
2011-04-24  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* ReceiverSocket.cc, cbcbsim.cc, 
	AIMD added as interest window control protocol.
2011-04-07  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* CBCBNode.cc, cbcbsim.cc, CCNxNode.cc,  
	rnd cache policy added
2011-02-04  Massimo GALLO Luca MUSCARIELLO <firstname.lastname@orange-ftgroup.com>
	* CCNxNodeCoding.cc, CCNxNodeCoding.h, CBCBNode.cc, cbcbsim.cc, ReceiverSocket.cc, ReceiverSocket.h
	Coding cache management option added
2011-02-04  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* CBCBNode.cc, CCNxNode.cc, CCNxNodeDiff.cc, DATA.h, INTEREST.h
	introduced file length (expressed in chunks) in the DATA and INTEREST packets
2011-02-04  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* CBCBNode.cc
	DRR is now done using name address + port (avoid to use same queue for two different content downloads expecially in the case of the 		   	   first link in which no cache and no filtering are used)
2011-02-18  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* CBCBNode.cc
	added the possibility to have infinite bandwidth and transfer packet based on the propagation delay (not yet coded in a proper way)
2011-02-17  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* many different files
	Diff and undiff simulators merged (possibility to manage cache for different applications)
	A file that specify the names that need statistics (option -classes k filename will do statistics chunk by chunk) 
        alternatively use -classes c number_of_classes that keeps statistics with class granularity for classes less or equal to the specified number
2010-11-16  Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CBCBNode.cc, include/CBCBNode.h, src/CCNxNode.cc, include/CCNxNode.h, include/CBCBNodeFactory.h, src/cbcbsim.cc
	Added possibility to specify different cache sizes for different nodes 
2010-10-26 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc include/CCNxNode.h
	added in stderr file the notion of different files downloading from a node
2010-10-06 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CBCBNode.cc src/CCNxNode.cc src/cbcbsim.cc
	removed possibility to cache and filter within the node that generates Interests (Client) 
2010-09-11 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/ReceiverSocket.cc
	temporal change to win protocol (no possibility to make FEC) and always resend interests not satisfied!
2010-06-11 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/cbcbsim.cc src/CBCBNode.cc
	added command option to disable filtering in client's caches
2010-06-11 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc cbcbsim/CCNxNode.h src/CBCBNode.cc cbcbsim/CBCBNode.h
	PIT lookup modified, added a parameter to disable filtering

2010-04-30 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc src/ReceiverSocket.cc
	Accept more contemporary identical downloading processes

2010-04-14 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc
	fixed a cache bug, accept cache size in kilobit

2010-04-13 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc
	improved structure of the Receiver socket list (exploit the new index: CCNx URI)

2010-04-12 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CCNxNode.cc
	improved search in the PIT_table (exploit the new index: CCNx URI)

2010-03-29 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CBCBDataCollector.cc, src/CBCBNode.cc, include/cbcbsim/CBCBDataCollector.h, include/cbcbsim/CBCBNode.h
	modified rate and loss calculation

2010-03-29 Massimo GALLO <firstname.lastname@orange-ftgroup.com>
	* src/CBCBNode.cc, src/DRR.cc, include/cbcbsim/DRR.h
	include the possibility of use DRR in the routers' queues


NodeFailWNP.cc
// -*- C++ -*-
//
//  This file is part of CBNSIM, the Content-Based Networking
//  Simulation Library.
//
//  Authors: See the file AUTHORS for full details.
//
//  Copyright (C) 2002-2004 University of Colorado
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
//  USA, or send email to serl@cs.colorado.edu.
//
// $Id: NodeFailWNP.cc,v 1.3 2004/04/19 05:26:16 rutherfo Exp $
//
#include <exception>
#include <siena/ssbg.h>
#include <siena/ssim.h>
#include <stdexcept>
#include "cbnsim/NodeFailWNP.h"
#include "cbnsim/logging.h"

using namespace cbnsim;
using namespace std;

NodeFailWNP :: NodeFailWNP( unsigned int nid,
			    ostream& out,
			    unsigned int delay,
			    unsigned int final_delay,
			    unsigned int recover_delay)
    : WorkloadNodeProcess( nid, out ), m_delay( delay ),
      m_final_delay( final_delay ), m_recover_delay( recover_delay )
{
    m_failed = false;
}

void NodeFailWNP :: init()
{
    if( m_final_delay )
    {
	m_final_time = ssbg::rand_poisson_delta( m_final_delay );
	ssim::Sim::self_signal_event( NULL, m_final_time );
    }
    
    if( m_delay )
    {
	ssim::Sim::self_signal_event( NULL, ssbg::rand_poisson_delta( m_delay ) );
    }
}

void NodeFailWNP :: print()
{
    if(m_failed)
    {
	m_out << "node_recover " << node_id() << " " << ssim::Sim::clock()
	      << " ;" << endl;
	m_failed = false;
    } else {
	m_out << "node_fail " << node_id() << " " << ssim::Sim::clock()
	      << " ;" << endl;
	m_failed = true;
    }
}

void NodeFailWNP :: process_event( const ssim::Event* evt )
{
  CBNASSERT( !evt, "NodeFailWNP can only handle NULL events" );
  if( m_final_time && ssim::Sim::clock() >  m_final_time )
    {
      m_failed = false;
      print();
    } 
  else
    {
      print();
      
      if( m_failed && m_recover_delay )
	{
	  ssim::Sim::self_signal_event( NULL, ssbg::rand_poisson_delta( m_recover_delay ) );
	}
      else if( !m_failed )
	{
	  ssim::Sim::self_signal_event( NULL, ssbg::rand_poisson_delta( m_delay ) );
	}
    }
}
